### DISCUSSION TITLE
How to "clean" upload folder?

### DISCUSSION THREAD

**Discussion by Gulpman** [immich-app/immich#10837] [Q&A]
I am new to immich and am uploading my pictures (via webbrowser on a PC) to immich via the upload function. I am doing this on Windows 11 and WSL2, as I have a huge number of pics to move (which I am doing step by step ;) ) and want to use the power of my graphics card there. When I am finished with the primary import I want to move the whole docker container + folders to my NAS.

When I first started importing I did not have Storage Templates enabled. I did so after having a look in the libary folder and found out the way my pics are stored there is useless for me as I didn't like the structure. I saw a few issues open here related to Storage Templates but as far as I could catch it this is only related when using the app - which I am not doing at the moment.

Now - after approx. 26000 pics have been uploaded I had a look inside the file structure.

My upload folder is containing 29 files, 25897 folders and takes up 571MB disk space.
I was wondering if:

- this is intended?
- if not (what I hope for ;) ): what would be the best way to remove this massive file structure?
  - can I just delete everything below E:\docker\immich\library\upload via the Windows Explorer?
  - is there a setting within immich which maybe has prevented that all the empty folders are deleted? Which would that be?
  - is there another safe way to get rid of this overhead?

Thank you very much in advance for your help.

Best
Sascha

**Comment by popy2k14** [comment]
+1
I am also stummbled on that folder during setup an rsync backup job.
Mine contains: 10 files, 50 022 folders, after uploading my complete library back then.

Here is a doc regarding backup: https://immich.app/docs/administration/backup-and-restore/
When i read this correvtly, you just have to backup the /library/ and exclude everything else.

- Can this please be confirmed?
- Can the massive structure be deleted?

I think a better way would be, that immich itself takes care of housekeeping/deleting the folders, when dont needed anymore!?
Should there be an ticket/issue created for this?

**Comment by mjessa12** [comment]
following because i also came across this - my upload folder is 23gb wondering if i can delete it

**Comment by lgrawet** [comment]
It depends if you are using storage template or not. The storage template mechanism is now disabled by default for new instances since [1.92.0](https://github.com/immich-app/immich/releases/tag/v1.92.0). This means files uploaded to the server will be stored in the upload folder and not moved to the library folder.

> Hardening storage template
>
>We have further hardened the storage template move mechanism to not leave any files in the upload folder behind. Making the job concurrency a fixed number and performing database lock between the move jobs will resolve some outstanding issues and stabilize this core feature.
>
>For new instances, the storage template mechanism is now disabled by default (existing instances won’t be affected). This means files uploaded to the server will be stored in the upload folder and not moved to the library folder. To provide a nice user experience, we added the onboarding feature, as seen below, to inform users that the storage template is disabled by default, and they can enable it before uploading any photos or videos to the instance to avoid confusion.
>
>This is also the continued effort to add support for other storage backends, such as Amazon S3 or Backblaze B2, in the future.

**Comment by popy2k14** [comment]
@lgrawet sorry i dont get it.
My understanding is, when i upload somthing, immich is using my storage template to sort and name the files accordingly.

So, does they get moved out of the upload folder or not?
If yes, i think i can safely delete the folder structure, right?
I think yes, because mine just have 10 files in it.

A confirmation would be nice :-)

Here is my template:
![image](https://github.com/user-attachments/assets/df5d7d77-095d-4bbd-91e0-20a8029a549e)

thx

**Reply by lgrawet** [reply]
Yes it should as you are using storage template. They had some problem in the past with leftovers though. That's one of the reasons that makes them move out from storage template.

**Comment by popy2k14** [comment]
Ok, thx for clearing things up.

So when I got you right, on current versions, there should be no leftover folders?

**Reply by lgrawet** [reply]
Not sure about that sorry. I don't think they are developing templating much further

**Reply by popy2k14** [reply]
ok, thx. i have cleaned it. We would see... :-)

**Reply by tobse91** [reply]
I am using the storage template and was wondering why there are a lot of empty folders and like 10 files in the uploaded folder, which, when using the storage template, should only be a temp folder for incoming uploads.
After deleting everything inside of it and uploading a new picture, new folders are created and are being left as empty folders.
Would be cool if immich would auto delete them after using. Otherwise it gets a bit confusing.

**Reply by zolakt** [reply]
The files do get deleted, but there are really a lot of these orphaned folder lying around. There are leftover folder in `/upload`, `/thumbs` and `/library`, probably in `/encoded-video` as well (haven't tried yet).

I know I'm nitpicking, but it bothers me to have hundreds of empty folders lying around. I'll probably write some script that runs periodically and removes all empty folders. But it would really be nice if such a script was baked in.

Basically just running `find ./ -depth -type d -empty -delete` inside the `UPLOAD_LOCATION` folder would do it

**Comment by DeltaTango69** [comment]
I have also storage template activated. But the files are stored in the library folder. But in the upload folder many subfolders are created, but these folders are all empty. I don't unterstand the sense of this.  

![image](https://github.com/user-attachments/assets/69afaed9-362b-4589-83a6-20575b442ebf)
![image](https://github.com/user-attachments/assets/467b7153-30ce-4093-803b-cb9b571642ba)
![image](https://github.com/user-attachments/assets/16205278-3b78-4b15-9e50-701bb8fd751d)

**Comment by nesalkiN** [comment]
Same. Started using Immich a couple of days ago and I use the storage template to move stuff to dirs that make sense.
253 dirs in my library and 20 962 empty dirs in upload.

**Comment by mescanne** [comment]
My preference would be for the storage template engine to remove empty folders when it moves files, or you can customize the number of folders in uploads created. For example, just a single folder for the first letter of the UUID of the filename would be sufficient for me (so at most 36 folders).

**Comment by baramofme** [comment]
Is there a way to forcefully convert and move the contents of the "uploaded" folder into the "/library/admin" subdirectory? The server statistics show that it's occupying 25GB, but in reality, including the "uploaded" folder, it's using 89GB.

I'm currently using a template feature, and I want to know exactly how much disk space images and videos are taking up based on the server statistics.

1. server statistics 
![image](https://github.com/user-attachments/assets/1d4a42ba-b2ee-4712-98cb-2a8711ded085)

2. actual disk usage by folder
```
root@immich:/opt/stacks/immich# tree --du -h | grep G]
[ 84G]  .
├── [ 84G]  library
│   ├── [ 25G]  library
│   │   └── [ 25G]  admin
│   │       └── [ 25G]  2024
│   ├── [1.1G]  thumbs
│   │   └── [1.1G]  ce7195b0-f160-4c8e-bb78-3c3c4a76382e
│   └── [ 58G]  upload
│       └── [ 58G]  ce7195b0-f160-4c8e-bb78-3c3c4a76382e
│           ├── [1.0G]  0c
│           │   ├── [1.0G]  02
│           │   │   └── [1.0G]  0c02d9df-394e-47df-91f1-7c98e0dbd195.mp4
│           ├── [1.3G]  15
│           ├── [1.7G]  37
│           │   ├── [1.4G]  43
│           │   │   └── [1.4G]  37439a21-2ac4-4e37-b7fc-9cdfd0cd92ea.mp4
│           ├── [1.5G]  3f
│           │   ├── [1.0G]  3d
│           │   │   └── [1.0G]  3f3d187f-9b4e-4dd9-9487-4acb91ef8d4a.mp4
│           ├── [1.3G]  58
│           │   └── [1.3G]  f7
│           │       └── [1.3G]  58f75663-b575-499d-b1f8-f57c3ae71d5f.mp4
│           ├── [1.5G]  86
│           ├── [1.1G]  97
│           │   ├── [1.0G]  3e
│           │   │   └── [1.0G]  973ea7d8-4421-43a7-ab24-d1dc47ee07ee.mp4
│           ├── [1.6G]  a5
│           ├── [1.1G]  ad
│           ├── [1.6G]  ae
│           ├── [1.2G]  be
│           ├── [1.0G]  c8
│           ├── [1.2G]  ee
│           ├── [1.0G]  fa
```

**Comment by kiwijunglist** [comment]
I also have this issue.  I have a single user for immich.  I use storage templates.

upload folder: 8.8GB | 105 files and 47,758 sub-folders.
- immich/upload/xxx-xxx-xxx-xxx-xxx ; contains 51 files | 4.8 GB (mostly video files but there are a few images)
- the remainder of the files are in the two letter sub-folders

library folder: 422.7GB | 54,364 files in 3,398 sub-folders
- arrangement as expected

I don't know what I should do with the "upload" folder files.

**Reply by mescanne** [reply]
I agree 100% -- I intend to do this:

dry run:
find . -type d -empty -print0 | xargs -0 -n 1 echo rmdir

real:
find . -type d -empty -print0 | xargs -0 rmdir

But with 47,758 sub-folders that's an insane number of folders. I also find using a harddrive is extremely slow for metadata (empty directories), while my SSD is fast.

**Comment by BernardoGiordano** [comment]
The software should behave like Musicbrainz Picard: when it moves files from a destination to another (through a feature very similar to Immich's), it recursively deletes any folder that is empty after move.

**Comment by seyfidin** [comment]
Is it safe to delete all these empty folders when using storage template?

**Reply by zolakt** [reply]
I delete them with cron, didn't have any problems:
`0 4 * * * find /UPLOAD_LOCATION -depth -type d -empty -delete`

I'm using storage templates, and both internal and external libraries.
I'm running this on the whole `UPLOAD_LOCATION` folder, since it's not just the `upload/` folder that has these orphaned folders, there are more in `thumbs/`, `/library` etc

**Comment by Real-Gecko** [comment]
I've encountered the issue too, I've uploaded ~130K photos and ~11K videos from 3 users and majority of the files were moved to `library` and organized according to template, but few hundreds of files (485 MiB) are "stuck" in `upload` dir, rerunning jobs for applying storage template seem to do nothing, files are still in place. Any advice on how to fix?

**Comment by popy2k14** [comment]
I have move thos assets from the upload dir locally, upüloaded them with the webui and deleted the empty folders.

**Comment by rahatzamancse** [comment]
It's been one and a half years, but still there have been no official updates on how to clean up the uploads directory and the relevant PostgreSQL cleanup.

My upload directory has grown to 5GB in size (27,842 items) and still growing whenever an upload fails. These are the types of logs I get:

```
Query failed : {
  durationMs: 3.861997000873089,
  error: PostgresError: duplicate key value violates unique constraint "UQ_assets_owner_checksum"
      at ErrorResponse (/usr/src/app/server/node_modules/.pnpm/postgres@3.4.7/node_modules/postgres/cjs/src/connection.js:794:26)
      at handle (/usr/src/app/server/node_modules/.pnpm/postgres@3.4.7/node_modules/postgres/cjs/src/connection.js:480:6)
      at Socket.data (/usr/src/app/server/node_modules/.pnpm/postgres@3.4.7/node_modules/postgres/cjs/src/connection.js:315:9)
      at Socket.emit (node:events:518:28)
      at addChunk (node:internal/streams/readable:561:12)
      at readableAddChunkPushByteMode (node:internal/streams/readable:512:3)
      at Readable.push (node:internal/streams/readable:392:5)
      at TCP.onStreamRead (node:internal/stream_base_commons:189:23)
      at TCP.callbackTrampoline (node:internal/async_hooks:130:17) {
    severity_local: 'ERROR',
    severity: 'ERROR',
    code: '23505',
    detail: 'Key ("ownerId", checksum)=(f2e07331-da1f-4ef6-8b00-6053042f9114, \\x3fdc315eb6749ec5434a2f25af45565b27915fad) already exists.',
    schema_name: 'public',
    table_name: 'asset',
    constraint_name: 'UQ_assets_owner_checksum',
    file: 'nbtinsert.c',
    line: '663',
    routine: '_bt_check_unique'
  },
  sql: 'insert into "asset" ("ownerId", "libraryId", "checksum", "originalPath", "deviceAssetId", "deviceId", "fileCreatedAt", "fileModifiedAt", "localDateTime", "type", "isFavorite", "duration", "visibility", "originalFileName") values ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14) returning *',
  params: [
    'f2e07331-da1f-4ef6-8b00-6053042f9114',
    null,
    <Buffer 3f dc 31 5e b6 74 9e c5 43 4a 2f 25 af 45 56 5b 27 91 5f ad>,
    '/data/upload/f2e07331-da1f-4ef6-8b00-6053042f9114/cb/fd/cbfda064-0692-405a-8605-8635c92f4a2d.HEIC',
    'FED6AF3C-45C6-482A-8DF1-714D058B48E6/L0/001',
    '9f03bd7cc1d685886b02be82c65e60ca3d6f827dc6cebfd7c69c1c654d8ff2c9',
    2025-11-29T04:17:34.000Z,
    2025-11-29T04:17:34.000Z,
    2025-11-29T04:17:34.000Z,
    'IMAGE',
    false,
    '0',
    'timeline',
    'IMG_6206.HEIC'
  ]
}
```

**Comment by jojolll** [comment]
For me, using the find command has been an effective solution to clean up unused empty folders. This command:

`find . -depth -type d -empty -delete`

allowed me to delete hundreds of thousands of unnecessary empty directories safely. From my experience, it only removes empty folder trees, and I haven't encountered any issues or negative side effects after running it.

**Comment by eddeT** [comment]
I just enabled templates to have a nice folder structure and also found it confusing that there is no cleanup of the upload dir. It's not a problem if I don't look at it, but now that I know there's a heap of empty directories, it's hard to ignore.

**Comment by taylorthurlow** [comment]
I have the same problem, and FWIW it looks like all the images remaining in the uploads folder are images that appear to have been interrupted when uploading - almost entirely RAW camera images that are missing/blank for some portion of the image when viewing it from a file browser.

**Comment by electricmessiah** [comment]
It's supposed to clean it, but I have yet to see evidence of this actually happening. There needs to be a serious code review here because these issues strike at the heart of this app, and it's not reliable if it continues to have these types of failures.

**Comment by geloczi** [comment]
FYI, this is still a bug in v3.1.0
